Docs: Update Securing Kubernetes Guide (#2758)
pomerium/pomerium: docs/guides/kubernetes.md

# Securing Kubernetes

The following guide covers how to secure [Kubernetes] using Pomerium. This is achieved by:

> - creating a ClusterRoleBinding for a user,

Author: I don't think so. Notice how the [Kubernetes Docs](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#api-overview) uses italics only to introduce the term, but uses code formatting for `kubectl` in the same paragraph.

```

Apply the configuration with:

```bash
kubectl apply -f ./pomerium-k8s.yaml
```

:::

### User Permissions

1. To grant access to users within Kubernetes, you will need to configure role-based access control (**RBAC**) permissions. For example, consider the example below, `rbac-someuser.yaml`:

    ```yaml
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRoleBinding
    metadata:
      name: cluster-admin-crb
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: cluster-admin
    subjects:
      - apiGroup: rbac.authorization.k8s.io
        kind: User
        name: someuser@example.com
    ```

1. Apply the RBAC ClusterRoleBinding:

    ```bash
    kubectl apply -f rbac-someuser.yaml
    ```

Permissions can also be granted to groups the Pomerium user is a member of.

## Create a Route for the API server

This new route requires a kubernetes service account token. Our Helm chart creates one and makes it available at `/var/run/secrets/kubernetes.io/serviceaccount/token`.

1. Update your `pomerium-values.yaml` file with the following route:

    ```yaml
      routes:
        - from: https://k8s.localhost.pomerium.io

>           to: https://kubernetes

Reviewer: ```suggestion
          to: https://kubernetes.default.svc.cluster.local
```

Reviewer: Less prone to error if we specify the default namespace

          tls_skip_verify: true
          kubernetes_service_account_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
          policy:
            - allow:
                or:
                  - domain:
                      is: pomerium.com
    ```

    Change the policy to match your configuration.

1. Apply the updated values with Helm:

    ```bash
    helm upgrade --install pomerium pomerium/pomerium --values pomerium-values.yaml
    ```

    This will create a Pomerium route within kubernetes that is accessible from `*.localhost.pomerium.io`.

## Configure Kubectl

The [pomerium-cli] tool can be used by kubectl as an auth helper. Once configured, connections to the cluster will open a browser window to the Pomerium authenticate service and generate an authorization token that will be used for Kubernetes API calls.

To use `pomerium-cli` as an exec-credential provider, update your kubectl config:

   ```shell
   # Add Cluster
   kubectl config set-cluster via-pomerium --server=https://k8s.localhost.pomerium.io
   # Add Context
   kubectl config set-context via-pomerium --user=via-pomerium --cluster=via-pomerium
   # Add credentials command

>    kubectl config set-credentials via-pomerium --exec-command=pomerium-cli --exec-arg=k8s,exec-credential,https://k8s.localhost.pomerium.io

Reviewer: ```suggestion
   kubectl config set-credentials via-pomerium --exec-command=pomerium-cli --exec-arg=k8s,exec-credential,https://k8s.localhost.pomerium.io --exec-api-version=client.authentication.k8s.io/v1beta1
```

Reviewer: Ref - https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuration

    ```yaml
      routes:
        - from: https://k8s.localhost.pomerium.io
          to: https://kubernetes
          allow_spdy: true
          tls_skip_verify: true
          kubernetes_service_account_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
          policy:
            - allow:
                or:
                  - domain:
                      is: pomerium.com
    ```

    Change the policy to match your configuration.

1. Apply the updated values with Helm:

    ```bash
    helm upgrade --install pomerium pomerium/pomerium --values pomerium-values.yaml
    ```

    This will create a Pomerium route within kubernetes that is accessible from `*.localhost.pomerium.io`.

## Configure Kubectl

The [pomerium-cli] tool can be used by kubectl as an auth helper. Once configured, connections to the cluster will open a browser window to the Pomerium authenticate service and generate an authorization token that will be used for Kubernetes API calls.

In addition to the methods listed in the page linked above, `pomerium-cli` can be installed via go-get:

```bash
env GO111MODULE=on GOBIN=$HOME/bin go get github.com/pomerium/pomerium/cmd/pomerium-cli@master
```

> Make sure `$HOME/bin` is on your path.

Author: Outstanding questions:

1. Should this section still be included, and/or should it be moved to the [Downloads page section](https://pomerium.com/docs/releases.html#pomerium-cli) for pomerium-cli?
2. If yes to either, should it be adjusted to point at the latest release branch instead of `master`, then updated as new major releases are cut?

Reviewer: I think just link to download/install section.  This was probably before we had a fully baked process for cli install.

Author: ```suggestion
```

    namespace: default
```

Apply the configuration with:

```bash
kubectl apply -f ./pomerium-k8s.yaml
```

:::

### User Permissions

1. To grant access to users within Kubernetes, you will need to configure role-based access control (**RBAC**) permissions. For example, consider the example below, `rbac-someuser.yaml`:

    ```yaml
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRoleBinding
    metadata:
      name: cluster-admin-crb
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: cluster-admin
    subjects:
      - apiGroup: rbac.authorization.k8s.io
        kind: User
        name: someuser@example.com
    ```

1. Apply the RBAC ClusterRoleBinding:

    ```bash
    kubectl apply -f rbac-someuser.yaml
    ```

> Permissions can also be granted to groups the Pomerium user is a member of.

Author: How's this?
```suggestion
Permissions can also be granted to groups the Pomerium user is a member of. This allows you to set a single ClusterRoleBinding in Kubernetes and modify access from your IdP.
```

## Create a Route for the API server

This new route requires a kubernetes service account token. Our Helm chart creates one and makes it available at `/var/run/secrets/kubernetes.io/serviceaccount/token`.

1. Update your `pomerium-values.yaml` file with the following route:

    ```yaml
      routes:
        - from: https://k8s.localhost.pomerium.io
          to: https://kubernetes.default.svc.cluster.local
          allow_spdy: true
          tls_skip_verify: true
          kubernetes_service_account_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
          policy:
            - allow:
                or:
                  - domain:
                      is: pomerium.com
    ```

    Change the policy to match your configuration.

1. Apply the updated values with Helm:

    ```bash
    helm upgrade --install pomerium pomerium/pomerium --values pomerium-values.yaml
    ```

    This will create a Pomerium route within kubernetes that is accessible from `*.localhost.pomerium.io`.

## Configure Kubectl

> The [pomerium-cli] tool can be used by kubectl as an auth helper. Once configured, connections to the cluster will open a browser window to the Pomerium authenticate service and generate an authorization token that will be used for Kubernetes API calls.

Reviewer 2: ```suggestion
The [pomerium-cli] tool can be used by kubectl as an authentication helper. Once configured, connections to the cluster will open a browser window to the Pomerium authenticate service and generate an authorization token that will be used for Kubernetes API calls.
```

Author: CC @travisgroth for what term is more accepted in the k8s space. Google search results favor "auth":
![image](https://user-images.githubusercontent.com/2349184/144089980-1c693bb4-6d79-47b8-9f78-96d9db1d5c38.png)
![image](https://user-images.githubusercontent.com/2349184/144090036-28178ca3-d927-4656-95dc-1564fc7781bf.png)

Reviewer: I'm not sure either is strictly accurate.  Maybe best to rephrase a bit.  According to https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins, we're technically a credential plugin.

Reviewer: ```suggestion
The [pomerium-cli] tool can be used by kubectl as credential plugin. Once configured, connections to the cluster will open a browser window to the Pomerium authenticate service and generate an authentication token that will be used for Kubernetes API calls.
```
